components/dashboard/src/service/json-rpc-user-client.ts

```diff
@@ -33,16 +38,174 @@ export class JsonRpcUserClient implements PromiseClient<typeof UserService> {
+        if (!userId) {
+            throw new ApplicationError(ErrorCodes.BAD_REQUEST, "userId is required");
+        }
+
+        const current = await getGitpodService().server.getLoggedInUser();
+        const currentAdditionalData = { ...current.additionalData };
+        const update: UpdateCurrentUserArgs = {
+            id: userId,
+            fullName: current.fullName!,
+            additionalData: {
+                ...(currentAdditionalData || {}),
+            },
+        };
+
+        if (emailNotificationSettings) {
+            if (!update.additionalData.emailNotificationSettings) {
+                update.additionalData.emailNotificationSettings = {};
+            }
+            const { allowsChangelogMail, allowsDevxMail, allowsOnboardingMail } = emailNotificationSettings;
+            if (typeof allowsChangelogMail === "boolean") {
+                update.additionalData.emailNotificationSettings.allowsChangelogMail = allowsChangelogMail;
+            }
+            if (typeof allowsDevxMail === "boolean") {
+                update.additionalData.emailNotificationSettings.allowsDevXMail = allowsDevxMail;
+            }
+            if (typeof allowsOnboardingMail === "boolean") {
+                update.additionalData.emailNotificationSettings.allowsOnboardingMail = allowsOnboardingMail;
+            }
+        }
+
+        if (typeof acceptedPrivacyPolicyDate === "string") {
+            update.additionalData.profile = update.additionalData.profile || {};
+            update.additionalData.profile.acceptedPrivacyPolicyDate = acceptedPrivacyPolicyDate;
+        }
+
+        if (typeof dotfileRepo === "string") {
+            update.additionalData.dotfileRepo = dotfileRepo;
+        }
+
+        if (typeof name === "string" && !!name.trim()) {
+            update.fullName = name;
+        }
+
+        if (typeof emailAddress === "string" && !!emailAddress.trim()) {
+            update.additionalData.profile = update.additionalData.profile || {};
+            update.additionalData.profile.emailAddress = emailAddress.trim();
+        }
+
+        if (!!editorSettings) {
+            update.additionalData.ideSettings = converter.fromEditorReference(
+                new EditorReference({ ...editorSettings }),
+            );
+        }
+
+        if (!!profile) {
+            update.additionalData.profile = update.additionalData.profile || {};
+            const {
+                explorationReasons,
+                signupGoals,
+                acceptedPrivacyPolicyDate,
+                companyName,
+                companySize,
+                jobRole,
+                jobRoleOther,
+                lastUpdatedDetailsNudge,
+                onboardedTimestamp,
+                signupGoalsOther,
+                emailAddress,
+            } = profile;
+            if (!!emailAddress) {
+                update.additionalData.profile.emailAddress = emailAddress;
+            }
+            if (!!explorationReasons) {
```

Comment:
Should array check with `.length`? Since init value may go wih EmptyArray.

Reply:
Let me check that 👀

Reply:
Tried to address it with https://github.com/gitpod-io/gitpod/pull/19238/commits/9da5892ee6057d21cf4f542d325a59c620643d5d